src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,489 @@
+		return nil, nil, errNoDelegationUsage
+	}
+
+	// Extract the algorithm used to sign the DelegatedCredential from the end-entity (leaf) certificate
+	var sigAlgo SignatureScheme
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		pk := sk.Public().(*ecdsa.PublicKey)
+		curveName := pk.Curve.Params().Name
+		certAlg := cert.Leaf.SignatureAlgorithm
+		if certAlg == x509.ECDSAWithSHA256 && curveName == "P-256" {
+			sigAlgo = ECDSAWithP256AndSHA256
+		} else if certAlg == x509.ECDSAWithSHA384 && curveName == "P-384" {
+			sigAlgo = ECDSAWithP384AndSHA384
+		} else if certAlg == x509.ECDSAWithSHA512 && curveName == "P-521" {
+			sigAlgo = ECDSAWithP521AndSHA512
+		} else {
+			return nil, nil, fmt.Errorf("using curve %s for %s is not supported", curveName, cert.Leaf.SignatureAlgorithm)
+		}
+	default:
+		return nil, nil, fmt.Errorf("unsupported algorithm for delegated credential")
+	}
+
+	// Generate the Delegated Credential Key Pair based on the provided scheme
+	var sk crypto.PrivateKey
+	var pk crypto.PublicKey
+	switch pubAlgo {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		sk, err = ecdsa.GenerateKey(getCurve(pubAlgo), rand.Reader)
+		if err != nil {
+			return nil, nil, err
+		}
+		pk = sk.(*ecdsa.PrivateKey).Public()
+	default:
+		return nil, nil, fmt.Errorf("unsupported algorithm for delegated credential: 0x%04x", pubAlgo)
+	}
+
+	// Prepare the credential for digital signing.
+	hash := getHash(sigAlgo)
+	credential := &Credential{validTime, pubAlgo, pk}
+	values, err := prepareDelegation(hash, credential, cert.Leaf.Raw, sigAlgo, peer)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sig []byte
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		opts := crypto.SignerOpts(hash)
+		sig, err = sk.Sign(rand.Reader, values, opts)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, fmt.Errorf("unsupported key type for delegated credential")
+	}
+
+	return &DelegatedCredential{
+		Cred:      credential,
+		Algorithm: sigAlgo,
+		Signature: sig,
+	}, sk, nil
+}
+
+// Validate checks that the delegated credential is valid by checking that the
+// signature is valid, that the credential hasn't expired, and that the TTL is
+// valid. It also checks that certificate can be used for delegation.
+func (dc *DelegatedCredential) Validate(cert *x509.Certificate, peer string, now time.Time) bool {
+	if dc.IsExpired(cert.NotBefore, now) {
+		return false
+	}
+
+	if dc.InvalidTTL(cert.NotBefore, now) {
+		return false
+	}
+
+	hash := getHash(dc.Algorithm)
+	in, err := prepareDelegation(hash, dc.Cred, cert.Raw, dc.Algorithm, peer)
+	if err != nil {
+		return false
+	}
+
+	// TODO: needs more thought
+	//if !(dc.Cred.expCertVerfAlgo == ECDSAWithP256AndSHA256 && cert.SignatureAlgorithm == x509.ECDSAWithSHA256) {
+	//	return false, errors.New("tls: delegated credential is not valid")
+	//}
+
+	if !isValidForDelegation(cert) {
```

Comment:
Perhaps this should be checked before we prepare the DC signature input?

Reply:
Yes!